utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+                                                , (.=)
+                                                , FromJSON(..)
+                                                , ToJSON(..)
+                                                , Value(String)
+                                                , decode
+                                                , object
+                                                , pairs
+                                                , withObject
+                                                , withText
+                                                )
+import           Data.Aeson.Types               ( Parser
+                                                , parseFail
+                                                )
+import qualified Data.ByteString.Char8         as B
+import qualified Data.ByteString.Lazy.Char8    as BL
+import           Data.Text                      ( Text )
+import qualified Data.Text                     as T
+import qualified Data.Text.IO                  as T
+import           Data.Vector                    ( Vector )
+import qualified Data.Vector                   as V
+import           Foreign.Storable               ( Storable(sizeOf) )
+import           GHC.Generics                   ( Generic )
+import           System.FilePath                ( (</>) )
+import           System.IO.Temp                 ( withSystemTempDirectory )
+import           System.Process.Typed           ( proc
+                                                , readProcess
+                                                )
+
+import           Control.Applicative            ( liftA2 )
+import           Control.Monad                  ( join )
+import           Data.Function                  ( on )
+import           Data.List                      ( groupBy
+                                                , sortOn
+                                                )
+import           Data.Map                       ( Map )
+import qualified Data.Map                      as M
+import           Data.Maybe                     ( catMaybes
+                                                , fromMaybe
+                                                )
+import           Data.Word                      ( Word32 )
+
+data EntryPoint = EntryPoint
+  { name :: Text
+  , mode :: ShaderStage
+  }
+  deriving (Show, Generic, FromJSON, ToJSON)
+
+data Input = Input
+  { type'    :: VertexAttributeType
+  , name     :: Text
+  , location :: Int
+  , array    :: Maybe (Vector Int)
+  }
+  deriving (Show, Generic)
+
+instance FromJSON Input where
+  parseJSON = withObject "inputs" $ \v ->
+    Input <$> v .: "type" <*> v .: "name" <*> v .: "location" <*> v .:? "array"
+
+instance ToJSON Input where
+  toJSON (Input type' name location array) = object
+    ["type" .= type', "name" .= name, "location" .= location, "array" .= array]
+  toEncoding (Input type' name location array) = pairs
```

Comment:
A `toJSON` method implementation should be enough. This part isn't anywhere near performance-critical paths.